README.md
# timesync

Time synchronization between peers.

Usage scenarios:

- **master/slave**: Clients synchronize their time to that of a single server,
  via either HTTP requests or WebSockets.
- **peer-to-peer**: Clients are connected in a (dynamic) peer-to-peer network
  using WebRTC or WebSockets and must converge to a single, common time in the
  network.


# Install

Install via npm:

```
npm install timesync
```


# Usage

A timesync client can basically connect to one server or multiple peers,
and will synchronize it's time. The synchronized time can be retrieved via
the method `now()`, and the client can subscribe to events like `'change'`
and `'sync'`.

```js
// create a timesync instance
var ts = timesync({
  server: '...',  // either a single server,
  peers: [...]    // or multiple peers
});

// get notified on changes in the offset
ts.on('change', function (offset) {
  console.log('offset from system time:', offset, 'ms');
}

// get the synchronized time
console.log('now:', new Date(ts.now()));
```


# Example

Here a full usage example with express.js, showing both server and client side.
`timesync` has build-in support for requests over http and can be used with
express, a default http server, or other solutions. `timesync` can also be
used over other transports than http, for example using websockets or webrtc.
This is demonstrated in the [advanced examples](/examples/advanced).

More examples are available in the [/examples](/examples) folder.
Some of the examples use libraries like `express` or `socket.io`.
Before you can run these examples you will have to install these dependencies.

**server.js**

```js
var express = require('express');
var timesyncServer = require('timesync/server');

// create an express app
var port = 8081;
var app = express();
app.listen(port);
console.log('Server listening at http://localhost:' + port);

// serve static index.html
app.get('/', express.static(__dirname));

// handle timesync requests
app.use('/timesync', timesyncServer.requestHandler);
```

**index.html**

```html
<!DOCTYPE html>
<html>
<head>
  <!-- note: for support on older browsers, you will need to load es5-shim and es6-shim -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/es5-shim/4.0.5/es5-shim.min.js"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/es6-shim/0.23.0/es6-shim.min.js"></script>

  <script src="/timesync/timesync.js"></script>
</head>
<script>
  // create a timesync instance
  var ts = timesync.create({
    server: '/timesync',
    interval: 10000
  });

  // get notified on changes in the offset
  ts.on('change', function (offset) {
    document.write('changed offset: ' + offset + ' ms<br>');
  });

  // get synchronized time
  setInterval(function () {
    var now = new Date(ts.now());
    document.write('now: ' + now.toISOString() + ' ms<br>');
  }, 1000);
</script>
</html>
```


# API

## Client

### Construction

An instance of timesync is created as:

```js
var ts = timesync(options);
```

#### Options

The following options are available:

Name       | Type                   | Default    | Description
---------- | ---------------------- | ---------- | ----------------------------------------
`delay`    | `number`               | `1000`     | Delay in milliseconds between every request sent.
`interval` | `number` or `null`     | `3600000`  | Interval in milliseconds for running a synchronization. Defaults to 1 hour. Set to `null` to disable automatically running synchronizations (synchronize by calling `sync()`).
`now`      | `function`             | `Date.now` | Function returning the local system time.
`peers`    | `string[]` or `string` | `[]`       | Array or comma separated string with uri's or id's of the peers to synchronize with. Cannot be used in conjunction with option `server`.
`repeat`   | `number`               | `5`        | Number of times to do a request to every peer.
`server`   | `string`               | none       | Url of a single server in case of a master/slave configuration. Cannot be used in conjunction with option `peers`.
`timeout`  | `number`               | `10000`    | Timeout in milliseconds for requests to fail.

### Methods

Name                  | Return type | Description
--------------------- | ----------- | ----------------------------------
`destroy()`           | none        | Destroy the timesync instance. Stops automatic synchronization. If timesync is currently executing a synchronization, this synchronization will be finished first.
`now()`               | `number`    | Get the synchronized time. Returns a timestamp. To create a `Date`, call `new Date(time.now())`.
`on(event, callback)` | `Object`    | Register a callback handler for an event. Returns the timesync instance. See section [Events](#events) for more information.
`off(event [, callback])` | `Object`    | Unregister a callback handler for an event. If no callback is provided, all callbacks of this event will be removed. Returns the timesync instance. See section [Events](#events) for more information.
`sync()`  | none        | Do a synchronization with all peers now.

To be able to send and receive messages from peers, `timesync` needs a transport. To hook up a transport like a websocket or http requests, one has to override the `send(id, data)` method of the `timesync` instance, and has to call `ts.receive(id, data)` on incoming messages.

Name                                | Return type | Description
----------------------------------- | ----------- | ----------------------------------
`send(to, data, timeout) : Promise` | none        | Send a message to a peer. `to` is the id of the peer, and `data` a JSON object containing the message. Must return a Promise which resolves when the message has been sent, or rejects when sending failed or a timeout occurred.
`receive(from, data)`               | none        | Receive a message from a peer. `from` is the id of the sender, and `data` a JSON object containing the message.

`timesync` sends messages using the JSON-RPC protocol, as described in the section [Protocol](#protocol).


### Events

`timesync` emits events when starting and finishing a synchronization, and when the time offset changes. To listen for events:

```js
ts.on('change', function (offset) {
  console.log('offset changed:', offset);
});
```

Available events:

Name     | Description
---------| ----------
`change` | Emitted when the offset is changed. This can only happen during a synchronization. Callbacks are called with the new offset (a number) as argument.
`error`  | Emitted when an error occurred. Callbacks are called with the error as argument.
`sync`   | Emitted when a synchronization is started or finished. Callback are called with a value `'start'` or `'end'` as argument.


### Properties

Name      | Type     | Description
--------- | -------- | --------------------------------------------
`offset`  | `number` | The offset from system time in milliseconds.
`options` | `Object` | An object holding all options of the timesync instance. One can safely adjust options like `peers` at any time. Not all options can be changed after construction, for example a changed `interval` value will not be applied.


## Server

`timesync` comes with a build in server to serve as a master for time synchronization. Clients can adjust their time to that of the server. The server basically just implements a POST request responding with its current time, and serves the static files `timesync.js` and `timesync.min.js` from the `/dist` folder. It's quite easy to implement this request handler yourself, as is demonstrated in the [advanced examples](/examples/advanced).

The protocol used by the server is described in the section [Protocol](#protocol).

### Load

The server can be loaded in node.js as:

```js
var timesyncServer = require('timesync/server');
```

### Methods

Name                          | Return type  | Description
----------------------------- | ------------ | ----------------------------------
`createServer()`              | `http.Server`| Create a new, dedicated http Server. This is just a shortcut for doing `http.createServer( timesyncServer.requestHandler )`.
`attachServer(server, [path])`| `http.Server`| Attach a request handler for time synchronization requests to an existing http Server. Argument `server` must be an instance of `http.Server`. Argument `path` is optional, and is `/timesync` by default.


### Properties

Name              | Type       | Description
----------------- | ---------- | --------------------------------------------
`requestHandler`  | `function` | A default request handler, handling requests for the timesync server. Signature is `requestHandler(request, response)`. This handler can be used to attach to an expressjs server, or to create a plain http server by doing `http.createServer( timesyncServer.requestHandler )`.


# Protocol

`timesync` sends messages using the JSON-RPC protocol. A peer sends a message:

```json
{"jsonrpc": "2.0", "id": "12345", "method": "timesync"}
```

The receiving peer replies with the same id and its current time:

```json
{"jsonrpc": "2.0", "id": "12345", "result": 1423151204595}
```

The sending peer matches the returned message by id and uses the result to adjust it's offset.


# Algorithm

`timesync` uses a simple synchronization protocol aimed at the gaming industry, and extends this for peer-to-peer networks. The algorithm is described [here](https://web.archive.org/web/20160310125700/http://mine-control.com/zack/timesync/timesync.html):

> A simple algorithm with these properties is as follows:
>
> 1. Client stamps current local time on a "time request" packet and sends to server
> 2. Upon receipt by server, server stamps server-time and returns
> 3. Upon receipt by client, client subtracts current time from sent time and divides by two to compute latency. It subtracts current time from server time to determine client-server time delta and adds in the half-latency to get the correct clock delta. (So far this algorithm is very similar to SNTP)
> 4. The first result should immediately be used to update the clock since it will get the local clock into at least the right ballpark (at least the right timezone!)
> 5. The client repeats steps 1 through 3 five or more times, pausing a few seconds each time. Other traffic may be allowed in the interim, but should be minimized for best results
> 6. The results of the packet receipts are accumulated and sorted in lowest-latency to highest-latency order. The median latency is determined by picking the mid-point sample from this ordered list.
> 7. All samples above approximately 1 standard-deviation from the median are discarded and the remaining samples are averaged using an arithmetic mean.

This algorithm assumes multiple clients synchronizing with a single server. In case of multiple peers, `timesync` will take the average offset of all peers (excluding itself) as offset.


# Tutorials

- [Using the timesync library in Android applications](https://github.com/enmasseio/timesync/blob/master/docs/android-tutorial.md)


# Resources

- [A Stream-based Time Synchronization Technique For Networked Computer Games](https://web.archive.org/web/20160310125700/http://mine-control.com/zack/timesync/timesync.html)
- [Network Time Protocol](http://www.wikiwand.com/en/Network_Time_Protocol)


# Build

To build the library:

    npm install
    npm run build

This will generate the files `timesync.js` and `timesync.min.js` in the folder `/dist`.

To automatically build on changes, run:

    npm run watch


HISTORY.md
# History


## 2022-09-05, version 1.0.11

- Fix #51: upgrade `debug` dependency to the latest version.


## 2021-10-24, version 1.0.10

- Fix #22: Emit error when the HTTP StatusCode is not 200 (see #44). 
  Thanks @gioid.


## 2021-10-24, version 1.0.9

- (oopsie, no changes merged, see 1.0.10)


## 2020-10-03, version 1.0.8

- Fix #31: add fields `"browser"` and `"react-native"` to package.json, 
  referring to the browser bundle `./dist/timesync.min.js`.


## 2020-06-29, version 1.0.7

- Fixed a try/catch using optional catch not supported by 
  all JavaScript engines. Thanks @cracker0dks.


## 2020-06-10, version 1.0.6

- Fix #31: library not working in React Native.
- Make the server side robust against requests with an invalid, 
  unparsable body: return ah HTTP 400 error. Thanks @calebTomlinson.
- Upgraded the dependency `debug` and all devDependencies.


## 2020-02-01, version 1.0.5

- Fix handing invalid JSON responses, see #27. Thanks @Poky85.


## 2019-10-12, version 1.0.4

- Move the `dependencies` only used in examples to `devDependencies`:
  `express`, `body-parser`, `socket.io`, promise`. To run the examples,
  you will have to install these dependencies yourself.


## 2018-05-21, version 1.0.3

- Fixed the `send` method passing a wrong parameter to `request.post`.


## 2018-03-07, version 1.0.2

- Fixed a broken example. Thanks @gokayokyay.


## 2018-02-12, version 1.0.1

- Upgraded some dependencies containing security vulnerabilities.


## 2018-02-12, version 1.0.0

- Function `send` now needs to return a Promise, and has to handle
  timeouts itself. This fixes timeouts occurring before xhr success.
  Thanks @DerekKeeler.
- Upgraded dependencies.


## 2017-10-23, version 0.4.2

- Fixed syncing to peers not using custom configuration `now`.
  Thanks @reklawnos.


## 2017-08-19, version 0.4.1

- Made examples robust against a varying working directory.
- Fixed #11: added timeout to http requests to fix infinite
  call / response loops. Thanks @dfeblowitz.


## 2017-06-13, version 0.4.0

- Added support for `https`. Thanks @jaquadro.
- Added a tutorial on how to use `timesync` in Android. Thanks @NicoHerbig.


## 2017-04-08, version 0.3.1

- Fixed a bug where JSON could be truncated on the server.
  Thanks Matt Kincaid.


## 2017-01-25, version 0.3.0

- Deliver ES5 version of the library in lib, bundled files in dist.
  Thanks @electrified.


## 2016-12-15, version 0.2.3

- Fixed some of the browser examples not working on Firefox.


## 2016-12-10, version 0.2.2

- Fixed library not working as part of a browserify setup. Thanks @SupremeTechnopriest.
- Fixed `sendTimestamp` not working in an express application. Thanks @SupremeTechnopriest.


## 2016-06-01, version 0.2.1

- Fixed #6: error when loading `timesync` in a node.js client.


## 2015-12-04, version 0.2.0

- Implemented an `'error'` event.
- Fixed unnecessary wait of 1 sec before the new offset is applied.
- Fixed #3: stat.median not using sorted array.


## 2015-02-17, version 0.1.0

- The first usable version.


## 2015-01-26, version 0.0.1

- Library name reserved at npm and bower.


server/index.js
var fs = require('fs');
var http = require('http');
var debug = require('debug')('timesync');

/**
 * Create a http server for time synchronization
 * @return {http.Server}      Returns the server instance.
 */
exports.createServer = function () {
  // create a new server
  debug('creating a new server');
  return http.createServer(exports.requestHandler);
};

/**
 * Attaches a timesync request hander to an existing server
 *
 * Implementation based on code from socket.io
 * https://github.com/Automattic/socket.io/blob/master/lib/index.js
 *
 * @param {http.Server} server    http server
 * @param {string} [path]         optional path, `/timesync` by default
 */
exports.attachServer = function (server, path) {
  if (server instanceof http.Server) {
    debug('attach to existing server');

    var listeners = server.listeners('request').slice(0);
    server.removeAllListeners('request');

    var route = createRoute(path);

    server.on('request', function(req, res) {
      var handled = route(req, res);
      if (!handled) {
        listeners.forEach(function (listener) {
          listener.call(server, req, res)
        });
      }
    });
  }
  else {
    throw new TypeError('Instance of http.Server expected');
  }
};

/**
 * Create a route matching whether to apply the timesync's request handler
 * @param {string} [path]         optional path, `/timesync` by default
 */
function createRoute (path) {
  path = path || '/timesync';
  var pattern = new RegExp(path + '($|/(.*))');

  debug('creating request handler listening on path ' + path);

  return function (req, res) {
    if (pattern.exec(req.url)) {
      exports.requestHandler(req, res);
      return true; // handled
    }
    else {
      return false; // not handled
    }
  };
}

/**
 * A request handler for time requests.
 * - In case of POST, an object containing the current timestamp will be
 *   returned.
 * - In case of GET .../timesync.js or GET .../timesync.min.js, the static
 *   files will be returned.
 * @param req
 * @param res
 * @returns {*}
 */
exports.requestHandler = function (req, res) {
  debug('request ' + req.method + ' ' + req.url + ' ' + req.method);

  if (req.method == 'POST') {
    if (!filename) {
      // a time request
      return sendTimestamp(req, res);
    }
  }

  if (req.method == 'GET') {
    var match = req.url.match(/\/(timesync(.min)?.js)$/);
    var filename = match && match[1];
    if (filename === 'timesync.js' || filename === 'timesync.min.js') {
      // static file handler
      return sendFile(res, __dirname + '/../dist/' + filename);
    }
  }

  res.writeHead(404);
  res.end('Not found');
};


function sendTimestamp(req, res) {
  if (req.body) {
    // express.js or similar
    sendTimeStampResponse(null, res, req.body);
  }
  else {
    // node.js
    readRequestBody(req, function (err, body) {
      sendTimeStampResponse(err, res, body);
    });
  }
}

function readRequestBody (req, callback) {
  var body = '';
  req.on('data', function (data) {
    body += data;

    // Too much POST data, kill the connection!
    if (body.length > 1e6) {
      callback(new Error('Too much post data. Killing connection...'), null);
      req.connection.destroy();
    }
  });
  req.on('end', function () {
    try {
      callback(null, JSON.parse(body));
    } catch(e) {
      callback(new Error('Invalid json in post body'));
    }
  });
}

function sendTimeStampResponse (err, res, body) {
  if (err) {
    res.writeHead(400);
    res.end(err.message);
  } else {

    var data = {
      id: 'id' in body ? body.id : null,
      result: Date.now()
    };

    // Set content-type header
    res.setHeader('Content-Type', 'application/json');

    if(res.status && res.send) {
      // express.js or similar
      res.status(200).send(JSON.stringify(data));
    } else {
      // node.js
      res.writeHead(200);
      res.end(JSON.stringify(data));
    }
  }

}


function sendFile(res, filename) {
  debug('serve static file ' + filename);
  fs.readFile(filename, function (err, data) {
    if (err) {
      res.writeHead(500);
      res.end('Error loading ' + filename.split('/').pop());
    }
    else {
      // note: content type depends on type of file,
      //       but in this case we only serve javascript files
      //       so we just hard code the Content-Type to application/javascript
      res.setHeader('Content-Type', 'application/javascript');
      res.writeHead(200);
      res.end(data);
    }
  });
}
